torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1424,6 +1433,458 @@ def rbinary_magic_impl(self, other):
+    # about it correctly. e.g. division depends on the denominator being
+    # different than 0.
+    #
+    # This class helps propagate such an assertion to the caller.
+    @dataclass
+    class Z3Node:
+        # Actual Z3 expression.
+        expr: Z3Expr
+        # Assertions for this expression.
+        assertion: Optional[z3.BoolRef] = None
+
+        @property
+        def arith(self) -> z3.ArithRef:
+            assert isinstance(self.expr, z3.ArithRef)
+            return self.expr
+
+        @property
+        def bool(self) -> z3.BoolRef:
+            assert isinstance(self.expr, z3.BoolRef)
+            return self.expr
+
+    # Implementation of Python semantics as Z3 expressions.
+    #
+    # Z3 Real-Int theory has operators with semantics that differ that of
+    # Python. Therefore, in order to get it right, we need to implement
+    # the (Python) semantics we are relying on in Z3.
+    class Z3Ops:
+        # The 2 functions below are used for conditionally casting between
+        # integer and reals.
+        #
+        # Returns a real expression from 'x'.
+        @staticmethod
+        def to_real(x: z3.ArithRef) -> Z3Node:
+            return Z3Node(x if x.is_real() else z3.ToReal(x))
+
+        # Returns an integer expression from 'x'.
+        @staticmethod
+        def to_int(x: z3.ArithRef) -> Z3Node:
+            return Z3Node(x if x.is_int() else z3.ToInt(x))
+
+        # Implements Python division semantics.
+        # This is needed because Z3 integer division does not have the
+        # same semantics as Python true division.
+        @staticmethod
+        def div(numerator: z3.ArithRef, denominator: z3.ArithRef) -> Z3Node:
+            denominator = Z3Ops.to_real(denominator).arith
+            return Z3Node(
+                expr=Z3Ops.to_real(numerator).arith / denominator,
+                assertion=_boolref(denominator != 0),
+            )
+
+        @staticmethod
+        def floor(number: z3.ArithRef) -> Z3Node:
+            return Z3Node(Z3Ops.to_int(number).expr)
+
+        @staticmethod
+        def floordiv(numerator: z3.ArithRef, denominator: z3.ArithRef) -> Z3Node:
+            div = Z3Ops.div(numerator, denominator)
+            return Z3Node(Z3Ops.to_int(div.arith).expr, div.assertion)
+
+        @staticmethod
+        def ceil(number: z3.ArithRef) -> Z3Node:
+            ite = z3.If(Z3Ops.floor(number).arith < number, Z3Ops.floor(number + 1).arith, number)
+            return Z3Node(cast(z3.ArithRef, ite))
+
+        @staticmethod
+        def mod(p: z3.ArithRef, q: z3.ArithRef) -> Z3Node:
+            q = Z3Ops.to_int(q).arith
+            return Z3Node(
+                expr=Z3Ops.to_int(p).arith % q,
+                assertion=_boolref(q != 0),
+            )
+
+        @staticmethod
+        def pow(base: z3.ArithRef, exp: z3.ArithRef) -> Z3Node:
+            return Z3Node(
+                expr=base ** exp,
+                assertion=_boolref(z3.Or(base != 0, exp > 0)),
```

Comment:
Blegh, if you do the guard thing I suggested for assertion, we'd have to start generating or guards in PyTorch. I suppose it's all feasible but these will be difficult to simplify away (unless we statically can discharge them)

Reply:
Hmm. Instead, we could just issue it to the validator, but outside this function (inside `make_node_magic` function).